### ISSUE TITLE
Switch from native-tls to rustls

### ISSUE THREAD

**Issue by autodidaddict** [nats-io/nats.rs#61]
As someone who has lost a cumulative total of 37 hours (yes, I've been keeping track) every time I've been stalled by a rust build chain that required the `native-tls` crate, which in turn requires the abomination known as `openssl-sys`, I would love for my use of the NATS client to not give its consumers this kind of headache.

Every time you build the `native-tls` crate in any kind of cross-platform scenario, you have to do so on a machine that has the open SSL source code/header libraries installed for the right target platform and OS and trust me when I say that even having this installed doesn't mean it'll work.

**tl;dr** - cross-platform/multi-platform build pipelines become insufferably difficult in the presence of anything that relies on openssl. The `rustls` crate does not have this problem.

**Comment by dwaynebradley** [comment]
@autodidaddict, I just remembered something after our Twitter conversations about this over the weekend. Even when I wanted to do a statically linked version of some Rust sample code with openssl, I had to "cross" compile it to `x86_64-unknown-linux-musl` and enable the following feature in my `cargo.toml` file:

```
[features]
# Force openssl-sys to staticlly link in the openssl library. Necessary when
# cross compiling to x86_64-unknown-linux-musl.
vendored = ["openssl-sys/vendored"]
```

I wonder if you would need to do something similar when cross compiling to armv7 or aarch64? Just a thought as I haven't tried that path.

I agree that moving to `rustls` may be much easier for things in the long run though.

**Comment by autodidaddict** [comment]
After a few hours, I stopped trying to cross compile, and just moved over to a native `aarch64` system, and _that_ system was still unable to properly locate the OpenSSL headers to compile.

**Comment by derekcollison** [comment]
So you could not even get it to compile on RPI with 64bit userland?

**Comment by autodidaddict** [comment]
I was eventually able to custom-build a Raspberry Pi 3B+ running 64-bit Ubuntu 20.04 LTS that _was_ able to find and compile the openssl-sys crate. However, the Pi just couldn't cope with the stack/memory requirements of the build and kept crashing. If I attempted the build (and cached the results from previous builds) about 6 times, I finally got the binary to generate :)

It might've been just fine on the 8GB RAM Pi4 running Ubuntu 20.04 LTS, though.

**Comment by dwaynebradley** [comment]
I just did a test using the [cross](https://github.com/rust-embedded/cross) project from the Rust Tools team on some sample Rust code with `nats.rs` with the feature like I mentioned above and that actually worked! I initially targeted armv7 as that is what I had available. Here is the command that I ran:

`cross build --target=armv7-unknown-linux-gnueabihf --features=openssl-sys/vendored`

**Comment by marcelbuesing** [comment]
While cross or the openssl-yss vendored flag is  a great quick fix to cross compile many projects I want to mention some disadvantage, compared to rustls, that i experienced:
- CI - I have not figured out how to directly use the docker images that cross uses internally directly in CI. So if you use cross in your CI you run docker in docker.
- You still require cross and can not just do "rustup target add ... && cargo build --target ..."
- Any library or project that depends on your library basically has the same issues, I'm frequently using `cargo tree -i -p openssl-sys` to identify what library is using openssl when my cross compiling fails.

I think a good approach is to give users the ability to switch between implementation via feature flags e.g.  [reqwest](https://github.com/seanmonstar/reqwest/blob/master/Cargo.toml#L36) . So while reqwest defaults to native-tls and you can opt out, I personally would default to rustls and allow switching to native-tls.

**Comment by spacejam** [comment]
This will be a relatively easy switch after @stjepang's async changes land. Currently we're using a bit of a hack to allow multiple concurrent threads to share a single TLS socket, but this hack will become unnecessary when the async core is merged. Rather than repurpose the hack, I'll wait for the new core to land where rustls support can be added in a cleaner way.

**Comment by dwaynebradley** [comment]
@marcelbuesing, completely agree with everything you mentioned. Using `cross` definitely isn't optimal or even workable sometimes. `rustls` would makes builds so much easier on everyone.

@spacejam / @stjepang, looking forward to seeing the update!

**Comment by stjepang** [comment]
Before switching to rustls, we need to make sure NATS is not using any of the technologies mentioned as "non-features" in rustl's README: https://github.com/ctz/rustls#non-features

@derekcollison Does anything there stand out?

**Comment by derekcollison** [comment]
Nope those are all things we avoid anyway.

**Comment by spacejam** [comment]
This will be published early next week in 0.7!
